### PR TITLE
[C++][Pistache] Generate API generalization interface 

### PR DESCRIPTION
**Pull request by CTerasa-ep**
The code generated for `main()` contains self-similar construction of, and calls to API implementation classes.

Example:
```
    PetApiImpl PetApiserver(router);
    PetApiserver.init();
    StoreApiImpl StoreApiserver(router);
    StoreApiserver.init();
    UserApiImpl UserApiserver(router);
    UserApiserver.init();
```

This sceams for OO generalization.

We could then store all `*server` objects inside one vector of API implementations and call `init()` on  all of them.
```
    auto apiImpls = std::vector<std::shared_ptr<ApiBase>>();

    apiImpls.push_back(std::make_shared<PetApiImpl>(router));
    apiImpls.push_back(std::make_shared<StoreApiImpl>(router));
    apiImpls.push_back(std::make_shared<UserApiImpl>(router));

    for (auto api : apiImpls) {
        api->init();
    }
```

This patch set refactors some Pistache generator code and introduces a base class for APIs called ApiBase and adds tehe above construction code.

Additionally do some refactoring work  in the Pistache generator code, that leverage some functional `stream()` methods that may make some intentions more explicit.

Technical Committee C++:
* @ravinikam 
* @stkrwork 
* @etherealjoy
* @martindelille
* @muttleyxd 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date.sh
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java [hunk 1]
```diff
@@ -267,61 +269,62 @@ public OperationsMap postProcessOperationsWithModels(OperationsMap objs, List<Mo
...
 
-        return objs;
+        // Check if any one of the operations needs a model, then at API file level, at least one model has to be included.
+        Predicate<String> importNotInImportMapping = hdr -> !importMapping.containsKey(hdr);
+        if (op.imports.stream().anyMatch(importNotInImportMapping)) {
+            operations.put("hasModelImport", true);
+        }
+    }
+
+    private static void postProcessSingleParam(CodegenParameter param) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:317]: thanks for the PR. 

Minor suggestion: what about adding a 1-liner docstring explaining what this function does?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java [hunk 2]
```diff
@@ -334,17 +337,19 @@ public String apiFilename(String templateName, String tag) {
...
         } else if (templateName.endsWith("impl-source.mustache")) {
-            int ix = result.lastIndexOf(File.separatorChar);
-            result = result.substring(0, ix) + result.substring(ix, result.length() - 4) + "Impl.cpp";
-            result = result.replace(apiFileFolder(), implFileFolder());
+            result = injectImplInFilename(result, ".cpp");
         }
         return result;
     }
 
+    private String injectImplInFilename(String filename, String suffix) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:347]: What about adding a 1-liner docstring (ideally with an example) explaining what this function does?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java (hunk 1 above)
Comment (CTerasa-ep) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:317]: Yep, I'll add the suggested docstring.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java (hunk 2 above)
Comment (CTerasa-ep) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:347]: Yep, I'll add the suggested docstring.

Comment (CTerasa-ep) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java]: Thinking about a comment I realized that he method could be named differently. Renamed the method to `implFilenameFromApiFilename`  in a follow-up commit 37c56049ccaa0a37ed2e9c421f4e5b27249ee439 and commented it. 
@wing328  Thank you, for the very valuable feedback!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java (hunk 1 above)
Comment (CTerasa-ep) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:322]: Documented with https://github.com/OpenAPITools/openapi-generator/pull/15279/commits/adf3e97e928f2ee7ddece576fbab9f6e8d21fd9c

**Review by muttleyxd — APPROVED** [review]

**Comment by muttleyxd** [comment]
C++ and template code looks good to me

**Comment by CTerasa-ep** [comment]
Great! Thank you for the feedback and your effort!

**Pull request merged**